images/router/haproxy/conf/haproxy-config.template

```diff
@@ -581,10 +583,10 @@ backend {{ genBackendNamePrefix $cfg.TLSTermination }}:{{ $cfgIdx }}
           {{- end }}
   tcp-request content reject if !whitelist
         {{- end }}
-        {{- with $value := clipHAProxyTimeoutValue (firstMatch $timeSpecPattern (index $cfg.Annotations "haproxy.router.openshift.io/timeout")) }}
+        {{- with $value := clipHAProxyTimeoutValue (firstMatch $timeSpecPattern (index $cfg.Annotations "haproxy.router.openshift.io/timeout") (env "ROUTER_DEFAULT_SERVER_TIMEOUT") "30s") }}
```

Comment:
All these additional `env` calls add some overhead.  As Andy says, "Trips across the Go/Template boundary can be expensive, particularly when you have many, many routes (10,000+)."  Consider adding `{{- $defaultServerTimeout := firstMatch $timeSpecPattern (env "ROUTER_DEFAULT_SERVER_TIMEOUT") "30s" }}` at the top of the template and using `{{- with $value := clipHAProxyTimeoutValue (firstMatch $timeSpecPattern (index $cfg.Annotations "haproxy.router.openshift.io/timeout") $defaultServerTimeout) }}` here.  

---

As noted in https://github.com/openshift/router/pull/483#pullrequestreview-1484951102, `clipHAProxyTimeoutValue (firstMatch $timeSpecPattern foo)` could be simplified as `clipHAProxyTimeoutValue foo`.  However, that might be better left as a separate change.  

...And actually, now that I look at it again, I believe `clipHAProxyTimeoutValue (firstMatch $timeSpecPattern) foo` will silently ignore invalid values while `clipHAProxyTimeoutValue foo` will log them.  Logging errors for invalid values might be a positive change, but it's an additional behavioral change that you might not want to tie to OCPBUGS-14914.

...Although thinking about it still further, I wonder:  Is `maxTimeoutFirstMatchedAndClipped` going to cause `clipHAProxyTimeoutValue` to log an error for each route that has an invalid annotation value, where before this change those invalid values get filtered out by `firstMatch` before `clipHAProxyTimeoutValue` has a chance to report any errors?

Reply:
>  Consider adding {{- $defaultServerTimeout := firstMatch $timeSpecPattern (env "ROUTER_DEFAULT_SERVER_TIMEOUT") "30s" }} at the top of the template and using {{- with $value := clipHAProxyTimeoutValue (firstMatch $timeSpecPattern (index $cfg.Annotations "haproxy.router.openshift.io/timeout") $defaultServerTimeout) }} here.

Done.